core/src/main/kotlin/com/strumenta/kolasu/ids/DeclarativeNodeIdProvider.kt

```diff
@@ -1,67 +1,145 @@
 package com.strumenta.kolasu.ids
 
 import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.validation.Issue
+import com.strumenta.kolasu.validation.IssueSeverity
+import com.strumenta.kolasu.validation.IssueSeverity.ERROR
+import com.strumenta.kolasu.validation.IssueSeverity.INFO
+import com.strumenta.kolasu.validation.IssueSeverity.WARNING
 import kotlin.reflect.KClass
-import kotlin.reflect.full.isSubclassOf
 import kotlin.reflect.full.isSuperclassOf
 
+// TODO merge into NodeIdProvider.kt file
+
 /**
- * Declarative node id provider instances allow to define
- * language-specific or custom identifier construction policies.
- *
- * ```kotlin
- * object MyDeclarativeNodeIdProvider: DeclarativeNodeIdProvider(
- *     idFor(PackageDeclaration::class) { it.node.name }
- *     idFor(ClassDeclaration::class) { (node: ClassDeclaration, identifiers: NodeIdProvider) ->
- *         val packageIdentifier = node.findAncestorOfType(PackageDeclaration::class)?.let { identifiers.idFor(it) } ?: "__none__"
- *         "${packageIdentifier}::${node.name}"
- *     }
- * )
- * ```
+ * Utility function to define a [NodeIdProvider] for a specific language.
+ * @param init the configuration rules for the [NodeIdProvider]
+ * @return a [NodeIdProvider] instance with the configured rules
  **/
-open class DeclarativeNodeIdProvider(
-    vararg rules: DeclarativeNodeIdProviderRule<out Node>
-) : NodeIdProvider {
-    private val rules: List<DeclarativeNodeIdProviderRule<out Node>> = rules.sorted()
-
-    override fun id(kNode: Node): String {
-        return this.rules.firstOrNull { it.canBeInvokedWith(kNode::class) }?.invoke(this, kNode)
-            ?: throw RuntimeException("Cannot find rule for node type: ${kNode::class.qualifiedName}")
-    }
+fun nodeIdProvider(
```

Comment:
Regarind NodeIdProviders, I have been doing some reflections today that I put in this document: https://docs.google.com/document/d/1beGyoVn6t8lrivs77_0P002xEXUGvXd3tCm1uZJQZJA/edit#heading=h.brcynsfenjhr

Reply:
Could not we have a concept of identifier in Kolasu which is not necessarily depending on LionWeb? We are using it to keep track of the correspondence between symbols and nodes, for example. Global symbol resolution should also be usable in context where information is kept in-memory - e.g. through an InMemorySymbolRepository - rather than LionWeb repositories. Could not LionWeb be just one possible option to store the information?

Reply:
I think that, for now, LionWeb is the only place where we need to use a node identifier. We may have later on situation where we may need a different way to handle identifiers but:
1. I do not know about such use cases
2. It may be the case that we can use the same logic to attribute identifiers that happen to work for LionWeb

So I would lean toward not generalizing at this stage

For the case where we use symbol repository in memory, wouldn't the same ID generated for LionWeb work too?

Reply:
> Ok, should we remove the DeclarativeNodeIdProvider and re-introduce it in the future in case we need it?

Yes

> The current symbol provision APIs work with a generic NodeIdProvider interface, so as long as the LionWeb StructuralNodeIdProvider works for in-memory cases everything should be fine.

Perfect, thank you